pkg/util/archive/zip.go

```diff
@@ -0,0 +1,261 @@
+
+	if !fileExists(destinationDir) {
+		err := os.MkdirAll(destinationDir, 0755)
+		if err != nil {
+			return fmt.Errorf("preparing destination: %v", err)
+		}
+	}
+
+	zipR, err := zip.OpenReader(source)
+	if err != nil {
+		return fmt.Errorf("opening source file: %v", err)
+	}
+	defer zipR.Close()
+
+	for _, f := range zipR.File {
+		err := extractAndWriteFile(f, destination)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extractAndWriteFile(f *zip.File, destination string) error {
+	rc, err := f.Open()
+	if err != nil {
+		return fmt.Errorf("failed to open file: %v", err)
+	}
+	defer rc.Close()
+
+	path := filepath.Join(destination, f.Name)
+
+	// Check for ZipSlip (Directory traversal)
+	if !strings.HasPrefix(path, filepath.Clean(destination)+string(os.PathSeparator)) {
+		return fmt.Errorf("illegal file path: %s", path)
+	}
+
+	if f.Mode() == os.ModeSymlink {
+		// We skip symlink for security reasons
+		return nil
+	}
+
+	if f.FileInfo().IsDir() {
+		err := os.MkdirAll(path, 0755)
+		if err != nil {
+			return fmt.Errorf("failed to create dir %s: %v", path, err)
+		}
+	} else {
+		err := os.MkdirAll(filepath.Dir(path), 0755)
+		if err != nil {
+			return fmt.Errorf("failed to file dir %s: %v", path, err)
+		}
+		f, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)
+		if err != nil {
+			return fmt.Errorf("failed to open file %s: %v", path, err)
+		}
+		defer f.Close()
+		_, err = io.Copy(f, rc)
+		if err != nil {
+			return fmt.Errorf("failed to copy file %s: %v", path, err)
+		}
+	}
+	return nil
+}
+
+// fileExists is an adapted implementation of fileExists from
+// https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L279
+// Copyright (c) 2016 Matthew Holt
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	return !errors.Is(err, fs.ErrNotExist)
+}
+
+// fileInfo is an adapted implementation of FileInfo from
+// https://github.com/mholt/archiver/blob/v3.5.1/archiver.go#L125
+// Copyright (c) 2016 Matthew Holt
+type fileInfo struct {
+	os.FileInfo
+	customName string
+}
+
+func (f fileInfo) Name() string {
+	if f.customName != "" {
+		return f.customName
+	}
+	return f.FileInfo.Name()
+}
+
+// writeWalk is an adapted implementation of (*Zip).writeWalk from
+// https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L300
+// Copyright (c) 2016 Matthew Holt
+func writeWalk(zipW *zip.Writer, source, destination string) error {
+	sourceInfo, err := os.Stat(source)
+	if err != nil {
+		return fmt.Errorf("%w stat: %s", err, source)
+	}
+	destAbs, err := filepath.Abs(destination)
+	if err != nil {
+		return fmt.Errorf("%w getting absolute path of destination %s: %s", err, destination, source)
+	}
+
+	return filepath.Walk(source, func(fpath string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("%w traversing %s", err, fpath)
+		}
+		if info == nil {
+			return fmt.Errorf("%s: no file info", fpath)
+		}
+
+		fpathAbs, err := filepath.Abs(fpath)
```

Comment:
I think it would be good to retain the [upstream comment](https://github.com/mholt/archiver/blob/v3.5.1/zip.go#L325-L326) here.